### PR TITLE
rgw/Zipper: DB Backend store

### PR DESCRIPTION
**Pull request by soumyakoduri**
This PR contains the initial bits of the DB backend (using SQLite) for Stackable RGW, part of project Zipper. The base class methods implemented are generic which could be extended to any backend database, not just SQLite.

More details on design/implementation can be found at - 
https://docs.google.com/document/d/1xCoHT5DCujqbe1pnEfYpSaSqiBcW8z87CgQCC5LivOo/edit#

**Current status:**
[Done]
    - User related APIs
    - Bucket APIs
    - Testcases (using Gtest) to test the DBStore APIs
    
[Not handled in the first pass]
    - Quota
    - Usage Stats
    - Swift Users
    - Multiple Zones/Zonegroups
    
[WIP]
    - Object APIs

### CODE REVIEW

**Comment by soumyakoduri** [comment]
> Please sign your commits. See also [1]. This is required for your PR to be merged.
> 
> [1] https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work
> 
> See [this article](https://help.github.com/en/articles/changing-a-commit-message) on GitHub on how to amend commits and update your pull request.

Done.

**Comment by soumyakoduri** [comment]
retest this please

**Comment by dang** [comment]
unstale

**Comment by soumyakoduri** [comment]
unstale

**Comment by soumyakoduri** [comment]
unstale

**Comment by soumyakoduri** [comment]
jenkins test this please

**Review by dang — APPROVED** [review]
Looks good.  Nice and clean.

File: src/rgw/rgw_sal_dbstore.h [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+    void encode(bufferlist& bl) const {
+      RGWOIDCProvider::encode(bl);
+    }
+    void decode(bufferlist::const_iterator& bl) {
+      RGWOIDCProvider::decode(bl);
+    }
+  };
+
+  class RGWDBStore : public Store {
```
[the remark this answers is not in the document]
Comment (soumyakoduri) [reply src/rgw/rgw_sal_dbstore.h]: Done.

**Review by alimaredia — APPROVED** [review]

**Comment by soumyakoduri** [comment]
https://pulpito.ceph.com/soumyakoduri-2021-07-16_16:33:05-rgw-dbstore-distro-basic-smithi/

**Comment by soumyakoduri** [comment]
http://pulpito.front.sepia.ceph.com/soumyakoduri-2021-07-28_14:23:01-rgw-dbstore-distro-basic-smithi/

File: src/rgw/CMakeLists.txt [hunk 2]
```diff
@@ -232,6 +232,10 @@ if(WITH_JAEGER)
   add_dependencies(rgw_common jaegertracing::libjaegertracing)
 endif()
 
+if(WITH_RADOSGW_DBSTORE)
+  target_link_libraries(rgw_common PRIVATE global dbstore)
```
Comment (soumyakoduri) [inline src/rgw/CMakeLists.txt:236]: @dang @kalebskeithley .. with commit#7b61667 , rgw_common is no more linked with rgw_a library and hence dbstore (rgw_sal_dbstore) had to be be linked directly to rgw_common now. 

Please let me know if dbstore should be linked to rgw_a target as well.

**Comment by soumyakoduri** [comment]
http://pulpito.front.sepia.ceph.com/soumyakoduri-2021-07-29_18:19:50-rgw-dbstore-distro-basic-smithi/

**Comment by soumyakoduri** [comment]
> I think this has complicated the repository. When cloning or performing other operations on macOS I get a collision with `FindSQLite3.cmake`. While Apple's quirky ideas around filename case are regrettable, I suggest that this file should be reconciled with the existing and different file whose name is casified differently.
> 
> ```
> Cloning into 'ceph'...
> remote: Enumerating objects: 1057640, done.
> remote: Counting objects: 100% (1100/1100), done.
> remote: Compressing objects: 100% (616/616), done.
> remote: Total 1057640 (delta 695), reused 737 (delta 478), pack-reused 1056540
> Receiving objects: 100% (1057640/1057640), 512.25 MiB | 21.29 MiB/s, done.
> Resolving deltas: 100% (844234/844234), done.
> Checking out files: 100% (10770/10770), done.
> warning: the following paths have collided (e.g. case-sensitive paths
> on a case-insensitive filesystem) and only one from the same
> colliding group is in the working tree:
> 
>   'cmake/modules/FindSQLITE3.cmake'
>   'cmake/modules/FindSQLite3.cmake'
> [lovecraft:~/git] aad%```
> ```

have opened PR (https://github.com/ceph/ceph/pull/42614 ) to fix this.

**Pull request merged**